src/operator/fusion/fused_op.cc

```diff
@@ -0,0 +1,269 @@
+ * under the License.
+ */
+
+#include "./fused_op.h"
+#include "../operator_common.h"
+#include "../../executor/exec_pass.h"
+
+#if MXNET_USE_CUDA
+
+namespace mxnet {
+
+DMLC_REGISTER_PARAMETER(FusedOpConfig);
+
+std::mutex FusedOp::mutex_;
+
+void FusedOpParamParser(nnvm::NodeAttrs* attrs) {
+  FusedOpConfig param;
+  try {
+    param.Init(attrs->dict);
+  } catch (const dmlc::ParamError& e) {
+    std::ostringstream os;
+    os << e.what();
+    os << ", in operator " << attrs->op->name << "("
+       << "name=\"" << attrs->name << "\"";
+    for (const auto& k : attrs->dict) {
+      os << ", " << k.first << "=\"" << k.second << "\"";
+    }
+    os << ")";
+    throw dmlc::ParamError(os.str());
+  }
+  attrs->parsed = FusedOpPtr(new FusedOp(attrs, param));
+}
+
+FusedOp::FusedOp(const nnvm::NodeAttrs* attrs, const FusedOpConfig& config) {
+  this->inputs_ = std::vector<FusedOpEntry>(config.num_inputs);
+  this->outputs_ = std::vector<FusedOpEntry>(config.num_outputs);
+  this->symbol_ = nnvm::Graph();
+  this->symbol_.outputs = attrs->subgraphs[0]->outputs;
+  this->initialized_ = false;
+  this->cc_major_ = -1;
+  this->cc_minor_ = -1;
+}
+
+bool FusedOp::InferShape(const nnvm::NodeAttrs &attrs,
+                         std::vector<mxnet::TShape> *in_attrs,
+                         std::vector<mxnet::TShape> *out_attrs) {
+  this->symbol_.attrs.erase("shape");
+  this->symbol_.attrs.erase("shape_inputs");
+  std::vector<mxnet::TShape> input_shapes(*in_attrs);
+  this->symbol_ = mxnet::exec::InferShape(std::move(this->symbol_),
+                                          std::move(input_shapes),
+                                          "__shape__");
+
+  const auto& g = this->symbol_.indexed_graph();
+  const auto& input_nids = g.input_nodes();
+
+  std::vector<mxnet::TShape> out_shapes;
+  const std::vector<mxnet::TShape> shapes = this->symbol_.GetAttr<mxnet::ShapeVector>("shape");
+  for (auto& e : g.outputs()) {
+    out_shapes.push_back(shapes[g.entry_id(e)]);
+  }
+  CHECK_EQ(out_shapes.size(), out_attrs->size());
+  for (size_t i = 0; i < out_attrs->size(); ++i) {
+    op::shape_assign(&(out_attrs->at(i)), out_shapes[i]);
+  }
+
+  // assign to in_attrs
+  for (size_t i = 0; i < in_attrs->size(); ++i) {
+    const auto eid = g.entry_id(input_nids[i], 0);
+    SHAPE_ASSIGN_CHECK(*in_attrs, i, shapes[eid]);
+  }
+
+  bool inferred = true;
+  for (const auto& attr : *in_attrs) {
+    inferred = inferred && !op::shape_is_none(attr);
+  }
+  for (const auto& attr : *out_attrs) {
+    inferred = inferred && !op::shape_is_none(attr);
+  }
+  return inferred;
+}
+
+bool FusedOp::InferType(const nnvm::NodeAttrs &attrs,
+                        std::vector<int> *in_attrs,
+                        std::vector<int> *out_attrs) {
+  this->symbol_.attrs.erase("dtype");
+  this->symbol_.attrs.erase("dtype_inputs");
+  std::vector<int> input_types(*in_attrs);
+  this->symbol_ = mxnet::exec::InferType(std::move(this->symbol_),
```

Comment:
I'm not sure if it would ever not be true, but ultimately it does not matter - if the outputs can be inferred that is enough. The only situation where the internal node's attributes are needed is backward, but then the main inferattr would fail due to backward node not knowing its attributes.